### PR TITLE
Add experimental causalml support

### PR DESCRIPTION
**Pull request by alexander-pv**
## Overview

Hi!
This PR is reopened https://github.com/shap/shap/pull/2654

## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

**Comment by connortann** [comment]
The tests are failing due to an unrelated HuggingFace error: apologies. We'll get that sorted asap (on #3286), and then we can get this PR passing.

Edit: done.

**Comment by connortann** [comment]
This PR has quite a large diff, as there are many unrelated style changes alongside the functional changes. Please would you remove the style changes, leaving just the functional changes?

We plan to introduce python black soon to remove this kind of hassle for contributors.

Nb. I inadvertently briefly closed this PR (currently on mobile), apologies.

**Comment by alexander-pv** [comment]
Apologies for style change. I will remove it.

Python black usage would be great!

**Comment by thatlittleboy** [comment]
Hi @alexander-pv , do you think you could also add in some tests to `test_tree.py` (just a simple function or two to demonstrate that the proposed changes here work as expected?)

Thanks!

**Comment by connortann** [comment]
I'll add the "stale" label for now as there hasn't been recent activity. Please feel free to comment if you're still interested in the PR and are happy to work on adding a minimal unit test.

**Comment by alexander-pv** [comment]
Hi, @connortann!  I added several test to check if shap works as intended with causalml trees.

**Comment by alexander-pv** [comment]
About current scikit-learn building issue in recent tests: https://github.com/scikit-learn/scikit-learn/issues/26858

**Comment by connortann** [comment]
Apologies that we haven't had time to look further at this PR, as many maintainers are away over the Christmas period. Thanks again for the PR - I hope to get to this early in the new year.

**Comment by alexander-pv** [comment]
Hi, @connortann! I updated the PR resolving a couple of conflicts. Tests work as expected on my computer.
So, awaiting feedback from label can be removed now. I will be glad to contribute if you think anything else is needed for this PR!

Thanks!

**Comment by jeongyoonlee** [comment]
Hi, @connortann. Can you take a look at this PR? This has been a long-standing issue (e.g., uber/causalml#265, uber/causalml#527, uber/causalml#615, uber/causalml#727, uber/causalml#735) for the `causalml` package, and we hope to resolve it soon.

Your support will be greatly appreciated.

**Comment by jeongyoonlee** [comment]
@alexander-pv, can you look into t[he test failures](https://github.com/shap/shap/actions/runs/10618029654/job/29432295181?pr=3273#step:6:797) in < Python3.12? Weirdly, it passes in Python 3.12.

File: tests/explainers/test_tree.py [hunk 1]
```diff
@@ -1641,6 +1642,64 @@ def test_lightgbm_call_explanation(self):
         assert (explanation.values == shap_values).all()
 
 
+class TestExplainerCausalML:
```
Comment (CloseChoice) [inline tests/explainers/test_tree.py]: we typically don't use classes in shap tests, would be great if you could refactor to use pytest.fixtures instead.

Also 
```python
        from causalml.dataset import synthetic_data
        from causalml.inference.tree import CausalTreeRegressor

        X, y, treatment = self.prepare_data(synthetic_data)
 ```       
could be handled in those. Pls also consider the usage of `pytest.importorskip` so that tests pass even if causalml is not installed.

**Review by CloseChoice — CHANGES_REQUESTED** [review]

**Comment by CloseChoice** [comment]
Thanks for the PR and sorry that this took so, so long. Will try to respond in a timely manner, pls feel free to ping me whenever you have questions or want another review.

**Comment by alexander-pv** [comment]
> Thanks for the PR and sorry that this took so, so long. Will try to respond in a timely manner, pls feel free to ping me whenever you have questions or want another review.

Ok, thanks!

File: tests/explainers/test_tree.py (hunk 1 above)
Comment (alexander-pv) [reply tests/explainers/test_tree.py]: Hi, @CloseChoice :wave: , can you review my latest changes, please?
Thanks

File: shap/explainers/_tree.py [hunk 2]
```diff
@@ -912,6 +912,7 @@ def __init__(self, model, data=None, data_missing=None, model_output=None):
                 "sklearn.ensemble.RandomForestRegressor",
                 "sklearn.ensemble.forest.RandomForestRegressor",
                 "econml.grf._base_grf.BaseGRF",
+                "causalml.inference.tree.CausalRandomForestRegressor",
```
Comment (CloseChoice) [inline shap/explainers/_tree.py:954]: is it possible that we add an info that causalml support is experimental using the logging module?

File: tests/explainers/test_tree.py [hunk 3]
```diff
@@ -1994,3 +1994,51 @@ def test_overflow_tree_path_dependent():
...
+
+    data, check_shape = causalml_synth_data
+    y, X, treatment, tau, b, e = data
+
+    ctree = CausalTreeRegressor(random_state=random_seed)
+    ctree.fit(X=X, treatment=treatment, y=y)
+    ctree_explainer = shap.TreeExplainer(ctree)
+    explanation = ctree_explainer(X)
+
+    shap_values_raw: list[np.ndarray] = ctree_explainer.shap_values(X)
```
Comment (CloseChoice) [inline tests/explainers/test_tree.py]: could you pls check the `__call__` interface as well?

File: tests/explainers/test_tree.py [hunk 4]
```diff
@@ -1994,3 +1994,51 @@ def test_overflow_tree_path_dependent():
...
+    shap_values: np.ndarray = np.stack(shap_values_raw, axis=-1)
```
Comment (CloseChoice) [inline tests/explainers/test_tree.py]: why is the output a list? For trees in general we moved to the shape
`[num_observations, ...features, n_classes]` for classifications and
`[num_observations, ...features, n_outputs]` for regression. Could you pls elaborate what each of these outputs mean? I couldn't find it at a quick glance at the causalml docs

**Review by CloseChoice — CHANGES_REQUESTED** [review]

File: shap/explainers/_tree.py (hunk 2 above)
Comment (alexander-pv) [reply shap/explainers/_tree.py:954]: Definitely :+1: TreeExplainer class constructor seems to be the best place for it. Is it OK to use warnings module as in the other functions of this module?

File: tests/explainers/test_tree.py (hunk 3 above)
Comment (alexander-pv) [reply tests/explainers/test_tree.py]: Ok :+1:

File: tests/explainers/test_tree.py (hunk 4 above)
Comment (alexander-pv) [reply tests/explainers/test_tree.py]: Thank you for the comment! Apologies, that's my mistake. It is in fact a numpy array. I'll fix it and add detailed explanation of the output.

File: shap/explainers/_tree.py (hunk 2 above)
Comment (CloseChoice) [reply shap/explainers/_tree.py:954]: yes, that's fine

File: tests/explainers/test_tree.py (hunk 4 above)
Comment (CloseChoice) [reply tests/explainers/test_tree.py]: First of all, I would never blame anyone here for any code contributed to shap (except for things that harm the project intentionally of course). Secondly we did the breaking change mid 2024, when this PR was already created and we just didn't check that due to capacity shortcomings and, I have to say, if anyone did a mistake here I would blame it on us, the maintainers. 

So long story short: thanks a lot for your contribution, we highly appreciate it.

File: shap/explainers/_tree.py (hunk 2 above)
Comment (alexander-pv) [reply shap/explainers/_tree.py:954]: Added `_safe_check_tree_instance_experimental` for that purpose

File: tests/explainers/test_tree.py (hunk 3 above)
Comment (alexander-pv) [reply tests/explainers/test_tree.py]: Added `__call__` output check ( `Explanation` class)

File: tests/explainers/test_tree.py (hunk 4 above)
Comment (alexander-pv) [reply tests/explainers/test_tree.py]: I think everything is correct now. Please, add comments if you notice something else during the review.
Thanks again for a quick feedback! :handshake:

File: shap/utils/_warnings.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class ExperimentalWarning(Warning):
```
Comment (alexander-pv) [inline shap/utils/_warnings.py:1]: Decided to suggest custom warnings similar to existing `_exceptions.py`. Of course, it is not very consistent to add a new module in this commit, but this class can clearly be reused during the initial integration of new models and I feel blue placing it inside `_tree.py` :sweat_smile: .
Let me know if it is better to avoid it at this time.

File: shap/explainers/_tree.py [hunk 6]
```diff
@@ -53,6 +54,41 @@
...
+    if hasattr(tree_instance, "__class__"):
+        if hasattr(tree_instance.__class__, "__module__"):
+            safe_instance = tree_instance
+
+    if safe_instance:
+        library = safe_instance.__class__.__module__.split(".")[0]
+        if experimental.get(library):
+            warnings.warn(
+                f"You are using experimental integration with {library}. "
+                f"The {library} support is verified for the following versions: {experimental.get(library)}.",
```
Comment (CloseChoice) [inline shap/explainers/_tree.py]: Can we make clear that we do not commit ourselves to supporting this long term
```suggestion
                f"The {library} support is verified for the following versions: {experimental.get(library)}. "
                 f"As experimental functionality, this integration may be removed or significantly changed in future releases without following semantic versioning. Use in production systems at your own risk.",
```

File: shap/explainers/_tree.py [hunk 7]
```diff
@@ -89,16 +125,16 @@ class TreeExplainer(Explainer):
...
-        model,
-        data=None,
-        model_output="raw",
-        feature_perturbation="auto",
-        feature_names=None,
-        approximate=DEPRECATED_APPROX,
-        # FIXME: The `link` and `linearize_link` arguments are ignored. GH #3513
-        link=None,
-        linearize_link=None,
+            self,
```
Comment (CloseChoice) [inline shap/explainers/_tree.py]: is there a particular reason for the formatting changes? I would to limit changes to either formatting or adding functionality, and not mix both into one PR.

File: shap/utils/_warnings.py (hunk 5 above)
Comment (CloseChoice) [reply shap/utils/_warnings.py:1]: I think that's a good idea, let's go with it.

File: tests/explainers/test_tree.py [hunk 8]
```diff
@@ -2000,45 +2001,96 @@ def test_causalml_causal_tree_explanation_output(causalml_synth_data, random_see
...
+
+        # Check Explanation class
+        assert explanation.data.shape == (n_observations, n_features)
+        assert explanation.base_values.shape == (n_observations, n_outcomes)
+        assert explanation.values.shape == (n_observations, n_features, n_outcomes)
+
+        # Check that shap values and base values can be collapsed into
+        # model prediction of individual treatment effects
+        y_outcomes = explanation.base_values + explanation.values.sum(axis=1)
+        individual_effects = y_outcomes[:, 1] - y_outcomes[:, 0]
```
Comment (CloseChoice) [inline tests/explainers/test_tree.py:2061]: If I understand it correctly the difference between the treatment effect and the non-treatment effect is the uplift, correct? This is exactly what we're calculating here, therefore it must match `preds`

**Review by CloseChoice — CHANGES_REQUESTED** [review]

**Comment by CloseChoice** [comment]
@alexander-pv I think we're almost through. Could you pls just remove the formatting changes (or explain why we need them) and add the comment regarding future support in the warning and we're good to merge.

File: tests/explainers/test_tree.py (hunk 8 above)
Comment (alexander-pv) [reply tests/explainers/test_tree.py:2061]: Yes, that's correct

File: shap/explainers/_tree.py (hunk 7 above)
Comment (alexander-pv) [reply shap/explainers/_tree.py]: Hmm, I didn't change formatting although it is showed in the merge commit https://github.com/shap/shap/pull/3273/commits/cc326cc3ecb9d655443af66ca69c0b1ab5701066. Anyway the latest diff shows that nothing changed.

**Comment by alexander-pv** [comment]
> @alexander-pv I think we're almost through. Could you pls just remove the formatting changes (or explain why we need them) and add the comment regarding future support in the warning and we're good to merge.

Done. Answered in the comments above.

**Review by CloseChoice — APPROVED** [review]
Very nice work. Thanks for the effort and patience; this has been a long time coming.

**Comment by CloseChoice** [comment]
@alexander-pv : I created a branch based on yours and fixed the pre-commit issues: https://github.com/shap/shap/tree/feature/causalml_trees. Pls merge that one in and then we'll merge once the tests pass

**Comment by alexander-pv** [comment]
> @alexander-pv : I created a branch based on yours and fixed the pre-commit issues: https://github.com/shap/shap/tree/feature/causalml_trees. Pls merge that one in and then we'll merge once the tests pass

Merged :ok_hand:
